File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        const forward = Array(nums.length).fill(1)
        const backwards = Array(nums.length).fill(1)
        for (let i =1; i < nums.length; i++) {
            forward[i] = nums[i-1] * forward[i-1]
        }
        for (let i = nums.length-2 ; i >=0; i--) {
            console.log(i, nums[i+1])
            backwards[i] = backwards[i+1] * nums[i+1]
        }
        const res = []
        for (let i = 0; i < nums.length; i++) {
            res.push(forward[i]*backwards[i])
        }
        return res
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(i: int, curr: List[int], total: int):
            if total == target:
                res.append(curr.copy())
                return
            
            if i >= len(nums) or total > target: 
                return
            
            # path 1: include current number
            curr.append(nums[i])
            dfs(i, curr, total + nums[i])

            # path2: don't include number
            curr.pop()
            dfs(i+1, curr, total)

        dfs(0,[],0)
        return res

